### PR TITLE
optimize(datastore): speed up opening datastore

### PR DESCRIPTION
**Pull request by jialeicui**
## Modules
- [ ] UI
- [ ] Controller
- [ ] Agent
- [x] Client
- [ ] Python-SDK
- [ ] Others

## Checklist
- [ ] run code format and lint check
- [ ] add unit test
- [ ] add necessary doc

### CODE REVIEW

File: client/starwhale/api/_impl/data_store.py [hunk 1]
```diff
@@ -1120,13 +1130,337 @@ def get_compressor(file: Path) -> Compressor:
...
+
+    def ahead(self, key: Any) -> bool:
+        return key is not None and self.max_key < key
+
+    def behind(self, key: Any) -> bool:
+        return key is not None and self.min_key > key
+
+    def __ge__(self, other: object) -> Any:
+        if not isinstance(other, DataBlock):
+            raise NotImplementedError
```
Comment (tianweidut) [inline client/starwhale/api/_impl/data_store.py:1244]: `return NotImplementedError`?

File: client/starwhale/api/_impl/data_store.py [hunk 2]
```diff
@@ -1120,13 +1130,337 @@ def get_compressor(file: Path) -> Compressor:
...
+    nodes: List[Node] = []
+    while len(iters) > 0 and len(nodes) == 0:
+        n = Node(iters.pop().iter)
+        if not n.exhausted:
+            nodes.append(n)
+
+    if len(nodes) == 0:
+        return
+
+    for i, iter in enumerate(iters):
```
Comment (tianweidut) [inline client/starwhale/api/_impl/data_store.py:1379]: iters is empty?

File: client/starwhale/api/_impl/data_store.py [hunk 3]
```diff
@@ -1120,13 +1130,337 @@ def get_compressor(file: Path) -> Compressor:
...
+        assert nodes[0].key is not None
+        if iter.min_key is None or nodes[0].key >= iter.min_key or len(nodes) == 0:
```
Comment (tianweidut) [inline client/starwhale/api/_impl/data_store.py:1381]: remove `len(nodes) == 0` condition?

File: client/starwhale/api/_impl/data_store.py [hunk 4]
```diff
@@ -1120,13 +1130,337 @@ def get_compressor(file: Path) -> Compressor:
...
+    while len(iters) > 0 and len(nodes) == 0:
+        n = Node(iters.pop().iter)
+        if not n.exhausted:
+            nodes.append(n)
+
+    if len(nodes) == 0:
+        return
+
+    for i, iter in enumerate(iters):
+        assert nodes[0].key is not None
```
Comment (tianweidut) [inline client/starwhale/api/_impl/data_store.py:1380]: mv `assert nodes[0].key is not None` outof the loop?

File: client/starwhale/api/_impl/data_store.py [hunk 5]
```diff
@@ -1120,13 +1130,337 @@ def get_compressor(file: Path) -> Compressor:
...
+        if iter.min_key is None or nodes[0].key >= iter.min_key or len(nodes) == 0:
+            n = Node(iter.iter)
+            iters.pop(i)
```
Comment (tianweidut) [inline client/starwhale/api/_impl/data_store.py:1383]: when some items was poped, the `i` still use the original index?

**Comment by tianweidut** [comment]
Others LGTM

File: client/starwhale/api/_impl/data_store.py [hunk 6]
```diff
@@ -42,8 +31,7 @@
 import requests
 import tenacity
 import jsonlines
-from filelock import FileLock
-from jsonlines import Writer
+from sortedcontainers import SortedDict, SortedList
```
[the remark this answers is not in the document]
Comment (jialeicui) [reply client/starwhale/api/_impl/data_store.py:34]: Done

File: client/starwhale/api/_impl/data_store.py (hunk 1 above)
Comment (jialeicui) [reply client/starwhale/api/_impl/data_store.py:1244]: All changed to `raise NotImplementedError`

File: client/starwhale/api/_impl/data_store.py (hunk 3 above)
Comment (jialeicui) [reply client/starwhale/api/_impl/data_store.py]: Changed to func

```py
def add_iters_to_node_if_needed(
    _nodes: List[Node], _iters: List[IterWithRangeHint]
) -> Tuple[List[Node], List[IterWithRangeHint]]:
    """
    Add the iterators to the nodes if the min key of the iterator is less than the global min key.
    This method have no side effect to the input parameters.
    """
    ret_nodes = _nodes[:]
    min_key = len(ret_nodes) > 0 and min(ret_nodes, key=lambda x: x.key).key or None
    filtered: List[IterWithRangeHint] = []
    for _iter in _iters:
        if (
            _iter.min_key is None
            or min_key is None
            or _iter.min_key <= min_key
            or len(ret_nodes) == 0
        ):
            _n = Node(_iter.iter)
            if not _n.exhausted:
                ret_nodes.append(_n)
                min_key = min(ret_nodes, key=lambda x: x.key).key
        else:
            filtered.append(_iter)
    return ret_nodes, sorted(filtered)
```

**Review by tianweidut — APPROVED** [review]

**Pull request merged**